uid-mux/src/yamux.rs

```diff
@@ -0,0 +1,440 @@
+            close_notify: Default::default(),
+            shutdown_notify: Default::default(),
+        }
+    }
+}
+
+impl<Io> IntoFuture for Yamux<Io>
+where
+    Io: AsyncWrite + AsyncRead + Unpin,
+{
+    type Output = Result<()>;
+    type IntoFuture = YamuxFuture<Io>;
+
+    fn into_future(self) -> Self::IntoFuture {
+        YamuxFuture {
+            role: self.role,
+            conn: self.conn,
+            incoming: Default::default(),
+            outgoing: Default::default(),
+            queue: self.queue,
+            closed: false,
+            remote_closed: false,
+            close_notify: self.close_notify,
+            shutdown_notify: self.shutdown_notify,
+        }
+    }
+}
+
+/// A yamux connection future.
+#[derive(Debug)]
+#[must_use = "futures do nothing unless you `.await` or poll them"]
+pub struct YamuxFuture<Io> {
+    role: Role,
+    conn: Connection<Io>,
+    /// Pending incoming streams, waiting for ids to be received.
+    incoming: FuturesUnordered<ReadId<Stream>>,
+    /// Pending outgoing streams, waiting to send ids and return streams
+    /// to callers.
+    outgoing: FuturesUnordered<ReturnStream<Stream>>,
+    queue: Arc<Mutex<Queue>>,
+    /// Whether this side has closed the connection.
+    closed: bool,
+    /// Whether the remote has closed the connection.
+    remote_closed: bool,
+    close_notify: Arc<Notify>,
+    shutdown_notify: Arc<AtomicBool>,
+}
+
+impl<Io> YamuxFuture<Io>
+where
+    Io: AsyncWrite + AsyncRead + Unpin,
+{
+    fn client_handle_inbound(&mut self, cx: &mut Context<'_>) -> Result<()> {
+        if let Poll::Ready(stream) = self.conn.poll_next_inbound(cx).map(Option::transpose)? {
+            if stream.is_some() {
+                error!("client mux received incoming stream");
+                return Err(
+                    std::io::Error::other("client mode can not accept incoming streams").into(),
+                );
+            }
+
+            info!("remote closed connection");
+            self.remote_closed = true;
+        }
+
+        Ok(())
+    }
+
+    fn client_handle_outbound(&mut self, cx: &mut Context<'_>) -> Result<()> {
+        let mut queue = self.queue.lock().unwrap();
+        while !queue.waiting.is_empty() {
+            if let Poll::Ready(stream) = self.conn.poll_new_outbound(cx)? {
+                let id = *queue.waiting.keys().next().unwrap();
+                let sender = queue.waiting.remove(&id).unwrap();
+
+                debug!("opened new stream: {}", id);
+
+                self.outgoing.push(ReturnStream::new(id, stream, sender));
+            } else {
+                break;
+            }
+        }
+
+        while let Poll::Ready(Some(())) =
+            self.outgoing.poll_next_unpin(cx).map(Option::transpose)?
+        {
+            trace!("finished processing stream");
+        }
+
+        // Set the waker to wake up when the queue is ready to process more streams.
+        queue.waker = cx.waker().clone();
```

Comment:
is there a reson why this is needed here? we will set the waker below with  self.set_queue_waker(cx);